File: libraries/python/examples/example_manager.py
"""Example: running multiple TRNSYS simulations with SimulationManager."""

from __future__ import annotations

import shutil
from pathlib import Path

from trnrun import SimulationConfig, SimulationManager

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
TRNEXE_PATH = Path(r"C:\TRNSYS18\Exe\TrnEXE64.exe")
MASTER_DCK = Path(r"examples\tpf\example_wo_plot_w_tracking.dck")
DCK_FOLDER = Path(r"examples\dck")

SIM_COUNT = 10
MAX_CONCURRENT = 5
MAX_PENDING = 3
REFRESH_INTERVAL = 1
CLEANUP_AFTER = True

CONFIG = SimulationConfig(trnexe_path=TRNEXE_PATH, watch_tmp=True)


# -----------------------------------------------------------------------------
# Helpers
# -----------------------------------------------------------------------------
def copy_dck(src: Path | str, dst_dir: Path | str, n: int) -> list[Path]:
    """Copy ``src`` into ``dst_dir`` ``n`` times with a zero-padded suffix."""
    src = Path(src)
    dst_dir = Path(dst_dir)
    _ = dst_dir.mkdir(parents=True, exist_ok=True)

    dst_files: list[Path] = []
    for i in range(1, n + 1):
        dst = dst_dir / f"{src.stem}_{i:03d}{src.suffix}"
        _ = shutil.copyfile(src, dst)
        dst_files.append(dst)
    return dst_files


def cleanup_folder(folder: Path | str) -> None:
    """Remove every file and subdirectory inside ``folder`` (folder is kept)."""
    folder = Path(folder)
    for item in folder.iterdir():
        if item.is_dir():
            shutil.rmtree(item)
        else:
            item.unlink()


# -----------------------------------------------------------------------------
# Run
# -----------------------------------------------------------------------------
def run_simulations(dck_files: list[Path]) -> SimulationManager:
    """Launch one simulation per deck and block until all have finished."""
    with SimulationManager(
        max_concurrent=MAX_CONCURRENT,
        max_pending=MAX_PENDING,
        refresh_interval=REFRESH_INTERVAL,
    ) as manager:
        for dck in dck_files:
            _ = manager.add(dck, CONFIG)
        _ = manager.wait()
    return manager


# -----------------------------------------------------------------------------
# Entry point
# -----------------------------------------------------------------------------
def main() -> None:
    """Create, submit, and optionally clean up example decks."""
    dck_files = copy_dck(MASTER_DCK, DCK_FOLDER, n=SIM_COUNT)
    _ = run_simulations(dck_files)

    if CLEANUP_AFTER:
        cleanup_folder(DCK_FOLDER)


if __name__ == "__main__":
    main()

File: libraries/matlab/toolbox/examples/example_manager.m
function simulations = example_manager()
%EXAMPLE_MANAGER Run multiple TRNSYS simulations with trnrun.SimulationManager.
%   SIMULATIONS = EXAMPLE_MANAGER() copies the bundled example deck several
%   times, submits every copy through one manager with bounded concurrency,
%   and returns the accepted simulations once they have all finished.

    example_folder = string(fileparts(mfilename("fullpath")));
    addpath(fileparts(example_folder))
    import trnrun.*

    % -------------------------------------------------------------------------
    % Configuration
    % -------------------------------------------------------------------------
    trnexe_path = "C:\TRNSYS18\Exe\TrnEXE64.exe";
    master_dck = fullfile(example_folder, "dck", "example_wo_plot_w_tracking.dck");
    dck_folder = fullfile(example_folder, "runs");

    sim_count = 20;
    max_concurrent = 10;
    refresh_interval = 1;

    config = SimulationConfig(trnexe_path=trnexe_path, watch_tmp=true);

    % -------------------------------------------------------------------------
    % Run
    % -------------------------------------------------------------------------
    dck_files = copy_dck(master_dck, dck_folder, sim_count);

    manager = SimulationManager(maxConcurrent=max_concurrent, refreshInterval=refresh_interval);

    for dck = dck_files
        manager.add(dck, config);
    end

    manager.wait();
    manager.shutdown();

    simulations = manager.simulations;

end

function dck_files = copy_dck(src, dst_dir, n)
%COPY_DCK Copy SRC into DST_DIR N times with a zero-padded suffix.

    if ~isfolder(dst_dir)
        mkdir(dst_dir);
    end

    [~, stem, extension] = fileparts(src);

    dck_files = strings(1, n);
    for index = 1:n
        dck_files(index) = fullfile(dst_dir, ...
            sprintf("%s_%03d%s", stem, index, extension));
        copyfile(src, dck_files(index));
    end
end
